### ISSUE TITLE
reportUnusedVariable does not complain about unused re-defined variables

### ISSUE THREAD

**Issue by Bruntaz** [microsoft/pyright#10102]
**Describe the bug**
I can't tell if this should be considered a bug or an enhancement request, sorry. The behaviour I'd like to change is that when using `reportUnusedVariable`, the final assignment of a variable should be disallowed if it is never used. 

**Code or Screenshots**

[Playground recreation](https://pyright-play.net/?code=MQAgDgngTglg5gCwC4C4RQKZgPZSQVQDsBXAZwwBMA1AQ1hoCMAbDAXiSmIwFgAoPihgBmIIdmwAKAJQo%2BIeSBohWIAIzzQAGQASAZRAxSIAMa5MxpEwghSCbAHdCio8RLkKfOQobKQAJi95Y18fAGo1DRAdfUMTMwwLKxs7R2cQVzJKT14FEApfAGZAvN988PUQLT0QbBEaUlJ4QgBbDEIkZOxiJnzbB3S3ShAANzoYRhYQBgSaTJAAfSQEQ3nnRrgWto7akAADCl2DF0GKIA).

```python
# pyright: reportUnusedVariable=true

def foo():
    a = 1  # LHS is correctly shown as unused

    b = 2
    c = b + 1  # LHS is correctly shown as unused

    d = 3
    d = d + 1  # LHS of assignment should show unused variable because _this_ assignment of `d` is unused
```

FWIW, this is how PyCharm's "unused variable" inspection works, which is why I'd expect Pyright's rule to behave like this (I'd like it implemented in Pyright for CI usage). Here's a screenshot from PyCharm **without** using Pyright's `reportUnusedVariable` option to show its behaviour: 

![Image](https://github.com/user-attachments/assets/29a48ad2-ef78-47ae-b4b0-f5ad57e03af6)

I've seen code issues before where entire chunks of code are pointless, but because they're assigned to a variable (which isn't used), linters have no issue with it. Enforcing that usages are required for redefined variables helps mitigate this issue a lot.

It's also worth noting that [Ruff can't do this (currently) because it requires branch analysis](https://github.com/astral-sh/ruff/issues/6704).

**VS Code extension or command-line**
Latest (1.1.394) as both language server and commandline.

**Comment by erictraut** [comment]
Pyright's `reportUnusedVariable` is working as intended, so this definitely isn't a bug.

What you're requesting here is very different from the check that `reportUnusedVariable` performs. It determines whether a variable that is assigned a value is consumed _anywhere_. It's implemented in the same way as the other `reportUnused...` checks.

What you're asking for is a check the determines whether a particular value assigned to a variable is subsequently used. 

This is a duplicate of enhancement request #9509. It has received a few upvotes, but it will require quite a few more for me to justify adding this feature. Even with many upvotes, I have concerns about the performance impact such a feature might impose because it requires a relatively expensive code graph walk for every variable assignment.
